src/producer/TransactionMQProducer.cpp

```diff
@@ -0,0 +1,218 @@
+  }
+
+  TransactionSendResult transactionSendResult(sendResult.getSendStatus(), sendResult.getMsgId(),
+                                              sendResult.getOffsetMsgId(), sendResult.getMessageQueue(),
+                                              sendResult.getQueueOffset());
+  transactionSendResult.setTransactionId(msg.getTransactionId());
+  transactionSendResult.setLocalTransactionState(localTransactionState);
+  return transactionSendResult;
+}
+
+void TransactionMQProducer::endTransaction(SendResult& sendResult, LocalTransactionState& localTransactionState) {
+  MQMessageId id;
+  if (sendResult.getOffsetMsgId() != "") {
+    id = MQDecoder::decodeMessageId(sendResult.getOffsetMsgId());
+  } else {
+    id = MQDecoder::decodeMessageId(sendResult.getMsgId());
+  }
+  string transId = sendResult.getTransactionId();
+
+  int commitOrRollback = 0;
+  switch (localTransactionState) {
+    case COMMIT_MESSAGE:
+      commitOrRollback = MessageSysFlag::TransactionCommitType;
+      break;
+    case ROLLBACK_MESSAGE:
+      commitOrRollback = MessageSysFlag::TransactionRollbackType;
+      break;
+    case UNKNOW:
+      commitOrRollback = MessageSysFlag::TransactionNotType;
+      break;
+    default:
+      commitOrRollback = MessageSysFlag::TransactionNotType;
+      break;
+  }
+
+  bool fromTransCheck = false;
+  EndTransactionRequestHeader* requestHeader =
+      new EndTransactionRequestHeader(getGroupName(), sendResult.getQueueOffset(), id.getOffset(), commitOrRollback,
+                                      fromTransCheck, sendResult.getMsgId(), transId);
+  LOG_DEBUG("endTransaction: msg:%s", requestHeader->toString().data());
+  getFactory()->endTransactionOneway(sendResult.getMessageQueue(), requestHeader, getSessionCredentials());
+}
+
+void TransactionMQProducer::checkTransactionState(const std::string& addr,
+                                                  const MQMessageExt& message,
+                                                  long tranStateTableOffset,
+                                                  long commitLogOffset,
+                                                  const std::string& msgId,
+                                                  const std::string& transactionId,
+                                                  const std::string& offsetMsgId) {
+                                                  
+  LOG_DEBUG("checkTransactionState: msgId:%s, transactionId:%s", msgId.data(), transactionId.data());
+  if (!m_transactionListener) {
+    LOG_WARN("checkTransactionState, transactionListener null");
+    THROW_MQEXCEPTION(MQClientException, "checkTransactionState, transactionListener null", -1);
+  }
+
+  m_ioService.post(boost::bind(&TransactionMQProducer::checkTransactionStateImpl, this, addr, message,
+                               tranStateTableOffset, commitLogOffset, msgId, transactionId, offsetMsgId));
+}
+
+void TransactionMQProducer::checkTransactionStateImpl(const std::string& addr,
```

Comment:
this method execute in another thread, don't pass reference as argument.

Reply:
it doesn't matter to use const reference